src/python/pants/backend/codegen/tasks/simple_codegen_task.py

```diff
@@ -172,3 +230,127 @@ def execute(self):
+
+    def __init__(self, task):
+      self._task = task
+
+    @property
+    def name(self):
+      return 'global'
+
+    def execute_codegen(self, targets):
+      self._task.execute_codegen(targets)
+
+    def find_sources(self, target):
+      """Predicts what sources the target will generate.
+
+      The exact implementation of this is left to the SimpleCodegenTask subclass.
+      :param Target target: the target for which to find generated sources.
+      :return OrderedSet: a set of relative filepaths.
+      """
+      return self._task.sources_generated_by_target(target)
+
+    def codegen_workdir_suffix(self, target):
+      return self.name
+
+
+  class IsolatedCodegenStrategy(CodegenStrategy):
+    """Code generate strategy which generates the code for each target separately.
+
+    Code is generated in a unique parent directory per target.
+    """
+
+    def __init__(self, task):
+      self._task = task
+      # NOTE(gm): This memoization yields a ~10% performance increase on my machine.
+      self._generated_sources_cache = {}
+
+    @property
+    def name(self):
+      return 'isolated'
+
+    def execute_codegen(self, targets):
+      for target in targets:
+        self._task.execute_codegen([target])
+
+    def find_sources(self, target):
+      """Determines what sources were generated by the target after the fact.
+
+      This is done by searching the directory where this target's code was generated. This is only
+      possible because each target has its own unique directory in this CodegenStrategy.
+      :param Target target: the target for which to find generated sources.
+      :return OrderedSet: a set of relative filepaths.
+      """
+      return self._find_sources_strictly_generated_by_target(target)
+
+    def codegen_workdir_suffix(self, target):
+      return os.path.join(self.name, target.id)
+
+    def _find_sources_generated_by_target(self, target):
+      if target.id in self._generated_sources_cache:
+        for source in self._generated_sources_cache[target.id]:
+          yield source
+        return
+      target_workdir = self._task.codegen_workdir(target)
+      if not os.path.exists(target_workdir):
+        return
+      for root, dirs, files in safe_walk(target_workdir):
+        for name in files:
+          yield os.path.join(root, name)
+
+    def _find_sources_generated_by_dependencies(self, target):
+      sources = OrderedSet()
+      def add_sources(dep):
+        if dep is not target:
+          dep_sources = self._find_sources_generated_by_target(dep)
+          dep_sources = [self._relative_source(dep, source) for source in dep_sources]
+          sources.update(dep_sources)
+      target.walk(add_sources)
+      return sources
+
+    def _relative_source(self, target, source):
+      return os.path.relpath(source, self._task.codegen_workdir(target))
+
+    def _find_sources_strictly_generated_by_target(self, target):
```

Comment:
There is something unsettling about this. Its implying that the isolated codegen is potentially generating the same source over and over again and then we're having to go back and filter it out.  I think that we have to do it, but it doesn't give me a warm fuzzy...